Fix CAM-13205

marcus-nl opened the pull request:
- don't attempt to remove decision requirements if DMN is disabled.
- extend existing unit test EmbeddedProcessApplicationTest.testDeployAppWithCustomEngine to reproduce the problem (without the above change it would *not* fail, but still log an exception).
- also fix a related bug in CmmnModelInstanceCache: getAllDefinitionsForDeployment erroneously returned DMN instead of CMMN definitions.
- additionally clarify error message in DeployProcessArchiveStep when the default process *engine* cannot be found.

koevskinikola commented:
Hi @marcus-nl,

Thank you for the contribution. I will review it in the following days and provide you with feedback. However, due to the upcoming code freeze for the 7.15 release, merging anything might take longer.

Best,
Nikola

koevskinikola commented:
Hi @marcus-nl,

I read the description in CAM-13205 and reviewed the code you provided.

First, I would recommend creating separate tickets and PRs for the bugs you are fixing:

* clarify error message in DeployProcessArchiveStep; 
* return CMMN instead of DMN definitions by the `CmmnModelInstanceCache#getAllDefinitionsForDeployment(..)` method

They fix separate issues that are unrelated.

Next, I wasn't able to reproduce the issue. According to your description, I should be able to see an exception in the logs when I execute the `EmbeddedProcessApplicationTest.testDeployAppWithCustomEngine()` test, with the `<property name="dmnEnabled">false</property>` set. However, I don't see any exceptions in the logs.

It would be better if you could provide a separate (new) test case/class without modifying the existing ones. This way, the issue would have a dedicated test. Currently, you're modifying a test case that covers a different functionality.

Best,
Nikola

koevskinikola reviewed (changes requested):
Please see the comment above.

marcus-nl commented:
Hi @koevskinikola,

I understand your desire for separate tickets / pull requests per issue. 

However, these issues are in fact related. Well, besides the error message clarification: that was just something that confused me while debugging, but TBH I'm not going to put effort into creating a ticket and PR just for changing one word ;). I realize that I didn't explain the relationship clearly in the ticket.

To clarify this the pull requests is now split into 3 commits:
1. A dedicated test to reproduce the issue. If you checkout this commit you should see [this stacktrace](https://gist.github.com/marcus-nl/d9df69988e74cc56d26cdef791233449) on undeploy.
2. The fix I proposed in the ticket: "The call to removeAllDecisionRequirementsDefinitionsByDeploymentId(..) should be moved up inside the `if (...isDmnEnabled())`" that was expected to fix the issue, but doesn't.
3.  The additional fix that is needed. The reason is that without this the CMMN-cache cleanup step (`cmmnModelInstanceCache.removeAllDefinitionsByDeploymentId(deploymentId);`) will actually query DMN definitions (which will fail if DMN is disabled) and attempt to remove those from its cache (which will never have any effect obviously). 

Also note the additional finding I wrote in the ticket:
"CMMN definitions were still removed from the cache, despite this bug! How is that possible? I did some debugging. As it turns out, the cache is also cleaned through DeploymentManager.deleteCaseDeployment (and deleteDecisionDeployment / deleteDecisionRequirementDeployment for DMN with the correct condition), called by deleteDeployment. This would suggest that DeploymentCache.removeDeployment is superfluous!"

So IMO you should consider removing DeploymentCache.removeDeployment entirely.

Regards, Marcus

koevskinikola commented:
Hi @marcus-nl,

Sorry for the long wait. I have now reviewed this PR. The code and your explanations are great, thank you!

I created a second PR where I can trigger our CI and test the changes (unfortunately, I can't trigger the CI on external contributor PRs). With commit 0951711, I added an additional assertion to the test you added, so that the presence of the exception is explicitly checked for.

If you're good with these changes, and the CI tests pass, I'll merge that PR instead of this one.

> This would suggest that `DeploymentCache.removeDeployment` is superfluous!"
>
> So IMO you should consider removing `DeploymentCache.removeDeployment` entirely.

Regarding your suggestion, I created CAM-13509 and will forward it for scheduling.

Best,
Nikola

marcus-nl commented:
Great, thanks!
